Makefile

```diff
@@ -0,0 +1,50 @@
+FEATURES_MINIMAL = --no-default-features --features 'std cargo' -p clap:3.0.0-beta.5
+FEATURES_FULL = --features 'wrap_help yaml regex unstable-replace unstable-multicall unstable-grouped'
+
+.PHONY: features-minimal features-full lint-minimal lint-full fmt-check lint coverage bench tests test debug clean
+
+features-minimal:
+	@echo "$(FEATURES_MINIMAL)"
+
+features-full:
+	@echo "$(FEATURES_FULL)"
+
+lint-minimal:
+	cargo clippy $(FEATURES_MINIMAL) -- -D warnings
+
+lint-full:
+	cargo clippy $(FEATURES_FULL) -- -D warnings
```

Comment:
What is the intention for all of the `make` logic?  CI just needs data, we could store it in a file, we aren't using any other make features and already there was concern in #2802 over using make.   If we aren't actually using make, I can't see this justify itself over the concerns of using make.

Reply:
Correction, I think it was only on this PRs conversation that concerns over make were raised.

Reply:
We have this makefile so that we can also show these commands in `CONTRIBUTING.md` instead of duplicating all these flags in there.

Reply:
And I am actually interested in that `make debug`. Was very helpful over the last few days.

Reply:
Whenever I contribute, I run the basic `cargo` workflow and then only respond to more when CI points it out.  As long as the CI lists the full command (which I confirmed this is doing), do we really need to document (and duplicate) everything the CI does?

Reply:
I found that it's best way to document a close replica of our CI process in `CONTRIBUTING.md`. Reduces feedback (CI) time for contributors and reduces the amount of CI runs our repository does relieving the runners free for other important runs.

Reply:
I've generally found CI is cheap, offload work to it rather than wasting time doing it locally.  I don't normally see people document contributing to this level or automate it.  Probably the closest is matklad's "do everything through `cargo test`" approach, which has its own trade offs.

Like I said in the other thread, is this really important enough?  We previously had a system for this (just) and, independent of this work, decided to get rid of it.

Reply:
I don't know about other projects in rust. But from my experience, all the big open source projects follow this rule of documenting the exact CI process in CONTRIBUTING.md

Reply:
My two cents as a potential contributor: leveraging CI in lieu of local testing is fine, but failing to document a procedure for locally verifying, at least, "CI will pass on my platform" is not. I personally prefer to iterate locally in most cases, and I like to be reasonably confident that I won't have to fix issues caught by CI after submitting a PR (or, failing that, not having to churn out noisy force-pushes to see if my fixes worked). Plus, it's reassuring to know exactly how to effectively test locally if CI falls over or is otherwise unavailable.

I didn't see anything wrong with the justfile in principle (I didn't have `just` installed, but it was in the Fedora repos so it wasn't hard to grab), but regardless, I think it's valuable to have a makefile or similar mechanism to quickly fire off a proper test run, rather than needing to copy-paste from the docs (or worse, CI configs) which I would be doing in any other case.

Reply:
In thinking of general principles, what I consider important are:
- Setting expectations (like "formatting is enforced" and "should pass under with various combination of feature flags')
- Meeting the user where they are
  - If I'm contributing to `clap`, its because I'm already using Rust and are more likely to be familiar with the ecosystem.  If I contribute to `bors`, it doesn't imply I use Elixir or am familiar with the Elixir ecosystem.  Level of documentation is proportionate to the expectations for contributors.
  - Emphasizing a user-driven workflow (their choice of editor, their choice of editor integration / command-line usage) rather than creating a bespoke workflow without there being a mitigating cause (e.g. setting up services in docker for end-to-end tests)
    - CI communicating what it does is a good bridge across workflows when responding to a failure
- Do not overwhelm the user with information such that they feel intimidated to contribute or can't find what they are looking for
- Easy to modify the process
  - Information isn't spread
  - Logic is straightforward, rather than going through hoops for the sake of DRY

Regardless, I am ok with going ahead and giving this a try to see how it works out, as long as it focuses on single sourcing CI information and documenting expectations.

Reply:
With that said, I really think we should emphasize calling into this from CI where possible to keep the Makefile from  going stale, like the just file.